### ISSUE TITLE
Outputs of a model are different depending on the batch size (tested for inception models)

### ISSUE THREAD

**Issue by jorgemf** [tensorflow/tensorflow#8670]
### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?

https://github.com/tensorflow/tensorflow/issues/1299

### Environment info
Operating System: ArchLinux
CUDA: 8.0.61
cuDNN: 5.1.10
TensorFlow: 1.0.1 (both with and without GPU support for python 2.7)

I  have also tested it with the latest nightly build:
http://ci.tensorflow.org/view/Nightly/job/nightly-matrix-cpu/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON2,label=cpu-slave/lastSuccessfulBuild/artifact/pip_test/whl/tensorflow-1.0.1-cp27-none-linux_x86_64.whl

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)

Minimun script for testing: https://gist.github.com/jorgemf/e1f1a97ad39c02ace3576d1248327d2a

I have tested it with the inception_v1 and inception_v4 models (commented in the code)

### What other attempted solutions have you tried?

None

### Logs or other output that would be helpful.

Logs for inception_v1 for the GPU and CPU based on my previous script.  You should expect the same output regardless of the batch size. Output must be the same for all the items of the batch as the input is the same image (which doesn't happen for one of the cases of this logs, I can upload the image and the checkpoint to reproduce that exactly case if required)

Please note that in this examples the errors in the output or the values of the layers are really small. You could think it is due to any weird conversion python does, but I have trained models based on inception_v1 where the outputs varies from 0.4 to 1.04.

The logs shows the output of the model (num_classes=1) and whether the layers outputs are exactly the same as with previous batch sizes or are different. Note that for some layers the outputs are the same regardless of the batch size, you start seeing errors with deeper layers. I only show few layers as example.

GPU
```
batch size: 1
        outputs: [  5.03433178e-13]
        OK inputs
        OK Conv2d_1a_7x7
        OK MaxPool_3a_3x3
        OK Mixed_3b
        OK Mixed_4b
        OK Mixed_4f
        OK Mixed_5c
batch size: 2
        outputs: [  5.03433233e-13]
        OK inputs
        OK Conv2d_1a_7x7
        OK MaxPool_3a_3x3
        OK Mixed_3b
        OK Mixed_4b
        ERROR Mixed_4f  Different values than batch_sizes 1.
        ERROR Mixed_5c  Different values than batch_sizes 1.
batch size: 10
        outputs: [  5.03433233e-13]
        OK inputs
        OK Conv2d_1a_7x7
        OK MaxPool_3a_3x3
        OK Mixed_3b
        OK Mixed_4b
        ERROR Mixed_4f  Different values than batch_sizes 1. Same values as batch_sizes 2.
        ERROR Mixed_5c  Different values than batch_sizes 1, 2.
batch size: 16
        outputs: [  5.03433233e-13]
        OK inputs
        OK Conv2d_1a_7x7
        OK MaxPool_3a_3x3
        OK Mixed_3b
        OK Mixed_4b
        ERROR Mixed_4f  Different values than batch_sizes 1, 2, 10.
        ERROR Mixed_5c  Different values than batch_sizes 1, 2, 10.
batch size: 20
        outputs: [  5.03433287e-13]
        OK inputs
        OK Conv2d_1a_7x7
        OK MaxPool_3a_3x3
        OK Mixed_3b
        OK Mixed_4b
        ERROR Mixed_4f  Different values than batch_sizes 1, 16. Same values as batch_sizes 2, 10.
        ERROR Mixed_5c  Different values than batch_sizes 1, 2, 16. Same values as batch_sizes 10.
```

CPU
```
batch size: 1
        outputs: [  5.03432962e-13]
        OK inputs
        OK Conv2d_1a_7x7
        OK MaxPool_3a_3x3
        OK Mixed_3b
        OK Mixed_4b
        OK Mixed_4f
        OK Mixed_5c
batch size: 2
        outputs: [  5.03433016e-13]
        OK inputs
        OK Conv2d_1a_7x7
        OK MaxPool_3a_3x3
        OK Mixed_3b
        OK Mixed_4b
        OK Mixed_4f
        ERROR Mixed_5c  Different values than batch_sizes 1.
batch size: 10
        outputs (ERROR: outputs must have the same value): [  5.03433178e-13   5.03433178e-13   5.03433178e-13   5.03433178e-13
   5.03433178e-13   5.03433178e-13   5.03433178e-13   5.03433178e-13
   5.03433016e-13   5.03433016e-13]
        OK inputs
        OK Conv2d_1a_7x7
        OK MaxPool_3a_3x3
        OK Mixed_3b
        OK Mixed_4b
        OK Mixed_4f
        ERROR Mixed_5c  Different values than batch_sizes 1. Same values as batch_sizes 2.
batch size: 16
        outputs: [  5.03433178e-13]
        OK inputs
        OK Conv2d_1a_7x7
        OK MaxPool_3a_3x3
        OK Mixed_3b
        OK Mixed_4b
        OK Mixed_4f
        ERROR Mixed_5c  Different values than batch_sizes 1. Same values as batch_sizes 2, 10.
batch size: 20
        outputs: [  5.03433178e-13]
        OK inputs
        OK Conv2d_1a_7x7
        OK MaxPool_3a_3x3
        OK Mixed_3b
        OK Mixed_4b
        OK Mixed_4f
        ERROR Mixed_5c  Different values than batch_sizes 1. Same values as batch_sizes 2, 10, 16.
```

**Comment by jorgemf** [comment]
I am not sure but I think this could be related with the full connected layers with a lot of inputs/outputs. Somehow the operations are not performed in the same order and the round error is different per batch size, this leads to an accumulation of the error different per batch size. When the layers are big (1024 neurons) the error adds and the output is completely different.

**Comment by skye** [comment]
Could you try this experiment with doubles instead of floats? I wonder if the extra precision will fix the stability issues. I'm not sure if this behavior is expected, maybe @jshlens can comment further.

**Comment by jorgemf** [comment]
@skye, I will test it but there will be still some error. What bothers me is that I would expect same error regardless the batch size.

I have tested more and I am seeing the first errors in `slim.conv2d(branch_2, 32, [3, 3], scope='Conv2d_0b_3x3')`. In most executions the initial error is there but sometimes there is no error there. I am testing always with the same mode loading the same checkpoint and with the same input.

**Comment by aselle** [comment]
To be clear
```
batch size: 1
        outputs: [  5.03433178e-13]
batch size: 2
        outputs: [  5.03433233e-13]
batch size: 10
        outputs: [  5.03433233e-13]
batch size: 16
        outputs: [  5.03433233e-13]
batch size: 20
        outputs: [  5.03433287e-13]
```
You are worried that these are not the same error? between batch size 1 and 2. Those numbers are practically the same near zero result.  The floating point operations used between different batches may in fact be different between different tensor sizes due to performance auto tuning. Do you have any concrete reason to worry about this? It looks like not a bug to me.

**Comment by jorgemf** [comment]
@aselle It is almost near 0 in that example, but I have a trained model that the error accumulates and it is close to 0.4 which it is a lot when the expected output is between 0 and 1, and what lead me to investigate further what was going on. My model is inception_v1 plus a couple of full connected layers.

I don't mind the floating precision point error. But I expect that error to be consistent. Sometimes I get error in the convolution layer, other times there is no error (with different batch sizes). Regardless any optimization the output of an operation must be the same with the same input, which it is not the case. For the same input I am getting different results in the operations.

In some cases I am also getting different outputs with a batch size, like the batch size is 20, I use the same input 20 times and get 3 or 4 different results in the outputs. It is not consistent.

Anyway, I did tweak my model to decrease that error and now I don't have the huge problem I had before. But I still think TensorFlow must return the same output for the same input, regardless of any optimization.

**Comment by aselle** [comment]
@jorgemf, I understand your desire for consistency, but almost every parallel system is non-deterministic or it accepts a significant speed compromise to ensure exact determinism. Your problem might be posed in a way that is on the edge of numeric stability. As such, a perturbation in order of operations caused by gpu or cpu parallelism scheduling could easily cause inconsistent results. The design tradeoff has been chosen to be as fast as possible even if numeric predictability is less clear.

If you want us to look at it more closely, please provide a reproducible test case that has a significant error than than the near zero errors of this test case. Thanks!

**Comment by jorgemf** [comment]
Closing the issue as I lost the model with the problem. I will open it again when I can replicate it.

**Comment by JackieLeeTHU11** [comment]
@aselle Hi  Aselle, I encounter a similar problem. I user inception_v2 as a base network. During training, the batchsize=128. During testing, if the batchsize=128, everything is ok. However, if the batchsize is smaller than 128 the results are different. And the precision declines as the batchsize drops. If the batchsize=1, the network will failed. I also used inception_v3, the same problems appered. However, if the base network is replaced with Alex network (tensorflow), everything goes well. I think I have not used slim or inception_v2 properly. Could you please give me some advices？

**Comment by JackieLeeTHU11** [comment]
@aselle Today, I replace the inception_v2 with vgg (slim), and everything goes well. So I think the bug is associated with inception_v1~v3. Do you has any reminder?

**Comment by jorgemf** [comment]
@JackieLeeTHU11 This is not related with the issue I opened. Please open your own issue and provide code examples of your problem.

**Comment by JackieLeeTHU11** [comment]
@jorgemf OK, I will. Thanks!
